tests/unit/multiclass/KNN_unittest.cc

```diff
@@ -226,3 +226,79 @@ TEST(KNN, classify_multiple_kdtree)
 
 
 }
+
+
+TEST(KNN, nearest_neighbours_brute)
+{
+	//simple non-trivial test
+	//initialize labels
+	SGVector<float64_t> lab({0.0, 0.0, 1.0, 1.0, 0.0});
+	//initialize features
+	SGMatrix<float64_t> feat({{2.0, 1.0}, {3.0, 1.0}, {6.0, 1.0}, {7.0, 1.0}, {3.5, 2.0}});
+	//divide training and testing parts
+	SGVector<index_t> train ({0, 1, 2, 3});
+	SGVector<index_t> test ({4});
+
+	auto labels = std::make_shared<MulticlassLabels>(lab);
+	auto features = std::make_shared<DenseFeatures<float64_t> >(feat);
+	auto features_test = features->clone()->as<DotFeatures>();
+	auto labels_test = labels->clone()->as<MulticlassLabels>();
+
+	int32_t k=3;
+	auto distance = std::make_shared<EuclideanDistance>();
+	auto knn=std::make_shared<KNN> (k, distance, labels, KNN_BRUTE);
+
+	features->add_subset(train);
+	labels->add_subset(train);
+	knn->train(features);
+
+	features_test->add_subset(test);
+	labels_test->add_subset(test);
+
+	auto dist = std::make_shared<EuclideanDistance>(features, features_test);
+	knn->set_distance(dist);
+
+	SGMatrix<index_t> NN = knn->nearest_neighbors();
+	
+	EXPECT_DOUBLE_EQ(feat(0, NN(0,0)), 3.0); EXPECT_DOUBLE_EQ(feat(1, NN(0,0)), 1.0);
+	EXPECT_DOUBLE_EQ(feat(0, NN(1,0)), 2.0); EXPECT_DOUBLE_EQ(feat(1, NN(1,0)), 1.0);
+	EXPECT_DOUBLE_EQ(feat(0, NN(2,0)), 6.0); EXPECT_DOUBLE_EQ(feat(1, NN(2,0)), 1.0);
+}
+
+
+TEST(KNN, nearest_neighbours_kdtree)
+{
+	//simple non-trivial test
+	//initialize labels
+	SGVector<float64_t> lab({0.0, 0.0, 1.0, 1.0, 0.0});
+	//initialize features
+	SGMatrix<float64_t> feat({{2.0, 1.0}, {3.0, 1.0}, {6.0, 1.0}, {7.0, 1.0}, {3.5, 2.0}});
+	//divide training and testing parts
+	SGVector<index_t> train ({0, 1, 2, 3});
+	SGVector<index_t> test ({4});
+
+	auto labels = std::make_shared<MulticlassLabels>(lab);
+	auto features = std::make_shared<DenseFeatures<float64_t> >(feat);
+	auto features_test = features->clone()->as<DotFeatures>();
+	auto labels_test = labels->clone()->as<MulticlassLabels>();
+
+	int32_t k=3;
+	auto distance = std::make_shared<EuclideanDistance>();
+	auto knn=std::make_shared<KNN> (k, distance, labels, KNN_KDTREE);
+
+	features->add_subset(train);
+	labels->add_subset(train);
+	knn->train(features);
+
+	features_test->add_subset(test);
+	labels_test->add_subset(test);
+
+	auto dist = std::make_shared<EuclideanDistance>(features, features_test);
+	knn->set_distance(dist);
+
+	SGMatrix<index_t> NN = knn->nearest_neighbors();
+	
+	EXPECT_DOUBLE_EQ(feat(0, NN(0,0)), 3.0); EXPECT_DOUBLE_EQ(feat(1, NN(0,0)), 1.0);
```

Comment:
picking up on our discussion, could you make this easier to read

* a comment might help
* fewer features might help
* it should be immediately obvious why you assert this

Reply:
is this fine?